pkg/schedulingtypes/jobscheduler.go

```diff
@@ -0,0 +1,229 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package schedulingtypes
+
+import (
+	"fmt"
+
+	"github.com/kubernetes-sigs/federation-v2/pkg/apis/core/typeconfig"
+	fedschedulingv1a1 "github.com/kubernetes-sigs/federation-v2/pkg/apis/scheduling/v1alpha1"
+	fedclientset "github.com/kubernetes-sigs/federation-v2/pkg/client/clientset/versioned"
+	"github.com/kubernetes-sigs/federation-v2/pkg/controller/util"
+	. "github.com/kubernetes-sigs/federation-v2/pkg/controller/util"
+	"github.com/kubernetes-sigs/federation-v2/pkg/controller/util/planner"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	pkgruntime "k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/apimachinery/pkg/watch"
+)
+
+const (
+	JSPKind          = "JobSchedulingPreference"
+	ParallelismField = "parallelism"
+	CompletionsField = "completions"
+	ParallelismPath  = "spec.parallelism"
+	CompletionsPath  = "spec.completions"
+)
+
+func init() {
+	schedulingType := SchedulingType{
+		Kind:             JSPKind,
+		SchedulerFactory: NewJobScheduler,
+	}
+	RegisterSchedulingType("jobs.batch", schedulingType)
+}
+
+type JobScheduler struct {
+	plugin *Plugin
+
+	fedClient fedclientset.Interface
+
+	controllerConfig *ControllerConfig
+
+	eventHandlers SchedulerEventHandlers
+}
+
+func NewJobScheduler(controllerConfig *ControllerConfig, eventHandlers SchedulerEventHandlers) (Scheduler, error) {
```

Comment:
(No action required) I'm not sure I understand the rational for an entirely new scheduler.  Plugins are already a thing - why can't a single controller delegate to type-specific plugins?